src/data_constructors.jl

```diff
@@ -89,35 +38,190 @@ and present inside `data`.
 """
 function data_random_effects(formula::FormulaTerm, data::D) where {D}
-    Tables.istable(data) || throw(ArgumentError("Data of type $D is not a table!"))
-    Z = nothing
-    # TODO:
-    # is_matrix_terms false for random-effects.
+    # with zerocorr we create only vectors and add them one by one with NCP
+    # without zerocorr we create a full-blown matrix with NCP
+    Z = Dict{String,AbstractArray}() # empty Dict
+    if !has_ranef(formula)
+        return nothing
+    end
+    slopes = slope_per_ranef(ranef(formula))
+
+    if length(slopes) > 0
+        # add the slopes to Z
+        # this would need to create a vector from the column of the X matrix from the
+        # slope term
+        for slope in values(slopes.grouping_vars)
+            if slope isa String
+                Z["slope_" * slope] = get_var(term(slope), data)
+            else
+                for s in slope
+                    Z["slope_" * s] = get_var(term(s), data)
+                end
+            end
+        end
+    else
+        Z = nothing
+    end
     return Z
 end
 
-function _has_ranef(formula::FormulaTerm, data)
-    return false
+"""
+    has_ranef(formula::FormulaTerm)
+
+Returns `true` if any of the terms in `formula` is a `FunctionTerm` or false
+otherwise.
+"""
+function has_ranef(formula::FormulaTerm)
+    return any(t -> t isa FunctionTerm, formula.rhs)
+end
+
+"""
+    ranef(formula::FormulaTerm)
+
+Returns a tuple of the `FunctionTerm`s parsed as `RandomEffectsTerm`s in `formula`.
+If there are no `FunctionTerm`s in `formula` returns `nothing`.
+"""
+function ranef(formula::FormulaTerm)
+    if has_ranef(formula)
+        terms = filter(t -> t isa FunctionTerm{typeof(|)}, formula.rhs)
+        terms = map(
+            t -> RandomEffectsTerm(first(t.args_parsed), last(t.args_parsed)), terms
+        )
+    else
+        terms = nothing
+    end
+    return terms
+end
+
+"""
+    n_ranef(formula::FormulaTerm)
+
+Returns the number of `RandomEffectsTerm`s in `formula`.
+"""
+function n_ranef(formula::FormulaTerm)
+    terms = ranef(formula)
+    if terms === nothing
+        return 0
+    elseif any(t -> t isa RandomEffectsTerm, terms)
+        counter = 0
+        for ts in terms
+            if ts.lhs isa Tuple
+                counter += tuple_length(ts.lhs)
+            else
+                counter += 1
+            end
+        end
+        return counter
+    else
+        # fallback
+        return 1
+    end
+end
+
+"""
+    intercept_per_ranef(terms::Tuple{RandomEffectsTerm})
+
+Returns a vector of `String`s where the entries are the grouping variables that have
+a group-level intercept.
+"""
+function intercept_per_ranef(terms::Tuple)
+    vec_intercepts = Vector{String}()
+    for ts in terms
+        if ts.lhs isa ConstantTerm
+            push!(vec_intercepts, string(ts.rhs))
+        elseif ts.lhs isa Tuple && any(t -> t isa ConstantTerm, ts.lhs)
+            push!(vec_intercepts, string(ts.rhs))
+        end
```

Comment:
Is this correct? The action for both conditionals is the same.

Reply:
Yes, I want it to error otherwise. Much more robust to bugs.

Reply:
That's not what I meant. You have written
```
if A
    do_something()
elseif B && C
    do_something()
end
```
That can be rewritten to
```
if A || (B && C)
    do_something()
end
```
Or any other variant, because the current way of writing the conditional is confusing.